pkg/kubectl/cmd/exec.go

```diff
@@ -174,16 +174,28 @@ func (p *ExecOptions) Run() error {
 		return err
 	}
 
-	if pod.Status.Phase != api.PodRunning {
-		return fmt.Errorf("pod %s is not running and cannot execute commands; current phase is %s", p.PodName, pod.Status.Phase)
-	}
-
 	containerName := p.ContainerName
 	if len(containerName) == 0 {
 		glog.V(4).Infof("defaulting container name to %s", pod.Spec.Containers[0].Name)
 		containerName = pod.Spec.Containers[0].Name
 	}
 
+	isInitContainer := false
+	for _, initContainer := range pod.Spec.InitContainers {
+		if initContainer.Name == containerName {
+			isInitContainer = true
+			break
+		}
+	}
+
+	if !isInitContainer && pod.Status.Phase != api.PodRunning {
```

Comment:
My previous comment applies here - handling this is the responsibility of the kubelet (returning an error back to the caller) not the client.